module/zfs/dmu_tx.c

```diff
@@ -1430,14 +1431,22 @@ dmu_tx_commit(dmu_tx_t *tx)
 void
 dmu_tx_abort(dmu_tx_t *tx)
 {
-	ASSERT(tx->tx_txg == 0);
+	/* This function should not be used on assigned transactions. */
+	ASSERT0(tx->tx_txg);
+
+	/* Should not be needed, but better be safe than sorry. */
+	if (tx->tx_tempreserve_cookie)
+		dsl_dir_tempreserve_clear(tx->tx_tempreserve_cookie, tx);
```

Comment:
Agreed, I cannot see how `tx_tempreserve_cookie`  could be set without a successful call to `dmu_tx_try_assign()`.

Reply:
Same as the other, if we didn't trip on assertion due to non-debug build.